src/ore/src/tracing.rs

```diff
@@ -463,19 +457,38 @@ where
     Ok((handle, guard))
 }
 
-/// Returns the level of a specific target from a [`Targets`].
-pub fn target_level(targets: &Targets, target: &str) -> Level {
-    if targets.would_enable(target, &Level::TRACE) {
-        Level::TRACE
-    } else if targets.would_enable(target, &Level::DEBUG) {
-        Level::DEBUG
-    } else if targets.would_enable(target, &Level::INFO) {
-        Level::INFO
-    } else if targets.would_enable(target, &Level::WARN) {
-        Level::WARN
-    } else {
-        Level::ERROR
+struct EmptyCallsite;
+impl Callsite for EmptyCallsite {
+    fn set_interest(&self, _: Interest) {
+        unreachable!("EmptyCallsite should never be accessed")
     }
+
+    fn metadata(&self) -> &Metadata<'_> {
+        unreachable!("EmptyCallsite should never be accessed")
+    }
+}
+
+/// Returns the level of a target from an [`EnvFilter`] by performing
+/// an exact match between `target` and the original `EnvFilter` directive
+/// (which may contain information like field names, span IDs, etc.)
+pub fn target_level(filter: &EnvFilter, target: &'static str) -> Level {
+    // EnvFilter roundtrips through its Display fmt, so it
+    // is safe to split out its individual directives here
+    for directive in format!("{}", filter).split(',') {
+        match directive.split('=').collect::<Vec<_>>().as_slice() {
+            [directive_target, "=", level] => {
+                if *directive_target == target {
+                    match Level::from_str(*level) {
+                        Ok(level) => return level,
+                        Err(err) => warn!("invalid level for {}: {}", target, err),
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    Level::ERROR
```

Comment:
Couple issues:
- This needs to default to the default level of the `EnvFilter`, not `ERROR` 
- It should probably ONLY support _top_level_ crates, as asking it for the `target_level` of `mz_storage::source` when `mz_storage=TRACE` would be incorrect. This can just be documented
- There needs to be a TODO to replace it with a `would_enable` impl on `EnvFilter` (I don't think this is untenable to implement upstream)

Reply:
> This needs to default to the default level of the EnvFilter, not ERROR

Hmm good point... `EnvFilter` doesn't have an explicit default level like `Targets` does, so that becomes kinda tricky. We don't have an easy way to root around its directives to see if there's a target-less level...